java/core/src/main/java/co/worklytics/psoxy/gateway/impl/CommonRequestHandler.java

```diff
@@ -122,17 +124,28 @@ public HttpEventResponse handle(HttpEventRequest request) {
             return healthCheckResponse.get();
         }
 
-        // re-write host
-        URL targetUrl = buildTarget(request);
-        String relativeURL = URLUtils.relativeURL(targetUrl);
+        // parse requested URL, re-writing host/etc
+        String requestedTargetUrl = parseRequestedTarget(request);
+
+        // if requested target URL has tokenized components, reverse
+        String clearTargetUrl = reverseTokenizedUrlComponents(requestedTargetUrl);
+
+        boolean tokenizedURLReversed = Objects.equals(requestedTargetUrl, clearTargetUrl);
+
+        String reversalOutcome = tokenizedURLReversed ? "No reversible tokens detected in URL." : "Detected tokens reversed.";
+
+        URL targetUrl = new URL(clearTargetUrl);
+
+        // avoid logging clear URL outside of dev
+        URL toLog = envVarsConfigService.isDevelopment() ? new URL(requestedTargetUrl) : targetUrl;
 
         boolean skipSanitization = skipSanitization(request);
 
         HttpEventResponse.HttpEventResponseBuilder builder = HttpEventResponse.builder();
 
         this.sanitizer = loadSanitizerRules();
 
-        String callLog = String.format("%s %s", request.getHttpMethod(), relativeURL);
+        String callLog = String.format("%s %s %s", request.getHttpMethod(), URLUtils.relativeURL(toLog), reversalOutcome);
```

Comment:
succinct boolean log, plus following my earlier comment log always the requested

```suggestion
String callLog = String.format("%s %s %s. Tokens reversed:%b", request.getHttpMethod(), URLUtils.relativeURL(requestedTargetUrl), tokenizedURLReversed);
```